File: design-patterns/python/State-DesignPattern/src/with_state_pattern/authenticated_state.py
"""Authenticated state - PIN correct, can withdraw."""
from .atm_state import ATMState
from .atm_machine import ATMMachine


class AuthenticatedState(ATMState):
    def insert_card(self, atm: ATMMachine) -> None:
        print("Card already inserted")

    def enter_pin(self, atm: ATMMachine, pin: int) -> None:
        print("Already authenticated")

    def withdraw_cash(self, atm: ATMMachine, amount: int) -> None:
        from .cash_dispensed_state import CashDispensedState
        if atm.get_balance() >= amount:
            print(f"Dispensing cash: {amount}")
            atm.set_balance(atm.get_balance() - amount)
            atm.set_state(CashDispensedState())
        else:
            print("Insufficient balance")

    def eject_card(self, atm: ATMMachine) -> None:
        from .idle_state import IdleState
        print("Card ejected")
        atm.set_state(IdleState())

File: design-patterns/java/State-DesignPattern/src/WithStatePattern/AuthenticatedState.java
package WithStatePattern;

public class AuthenticatedState implements ATMState {

    public void insertCard(ATMMachine atm) {
        System.out.println("Card already inserted");
    }

    public void enterPin(ATMMachine atm, int pin) {
        System.out.println("Already authenticated");
    }

    public void withdrawCash(ATMMachine atm, int amount) {

        if(atm.getBalance() >= amount){
            System.out.println("Dispensing cash: " + amount);
            atm.setBalance(atm.getBalance() - amount);
            atm.setState(new CashDispensedState());
        } else {
            System.out.println("Insufficient balance");
        }
    }

    public void ejectCard(ATMMachine atm) {
        System.out.println("Card ejected");
        atm.setState(new IdleState());
    }
}
